### DISCUSSION TITLE
Package with rollup

### DISCUSSION THREAD

**Discussion by hahaxiaowai** [visgl/deck.gl#5299] [Q&A]
I referenced deck.gl using npm, then I try to pack it with rollup, But it reported an error
``` node
[!] Error: 'global' is not exported by node_modules\probe.gl\env.js, imported by node_modules\@deck.gl\core\dist\esm\lib\init.js
https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module
node_modules\@deck.gl\core\dist\esm\lib\init.js (3:9)
1: import { registerLoaders } from '@loaders.gl/core';
2: import { ImageLoader } from '@loaders.gl/images';
3: import { global } from 'probe.gl/env';
            ^
4: import log from '../utils/log';
5: import { register } from '../debug';
```
This is an error when calling probe.gl in deck.gl，Because the variable Global is not exported by default.
I checked the [rollup document](https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module), it said it was an error caused by commonjs, and also gave a solution. Although this method can be packaged successfully, the file cannot be used normally after packaging.

**Comment by hahaxiaowai** [comment]
This is how i import
``` node
import {Tile3DLayer} from "@deck.gl/geo-layers";
import {MapboxLayer} from "@deck.gl/mapbox";
import {registerLoaders} from "@loaders.gl/core";
import {DracoWorkerLoader} from "@loaders.gl/draco";
import {CesiumIonLoader} from "@loaders.gl/3d-tiles";
```

**Comment by jeffpeck10x** [comment]
Were you able to figure out a solution to this?

Is it possible to bundle deck.gl with rollup?

**Reply by Pessimistress** [reply]
The code referenced by this thread is no longer present in the latest release (v8.7).

**Reply by jeffpeck10x** [reply]
Thanks! I just tried upgrading, but then had the same issue with current luma dependencies.

I hackily dealt with those like this:
```
find node_modules/@luma.gl -name "*.js" -path "*/dist/*" -exec sed -i ""  "s/'probe\.gl\/env'/'@probe.gl\/env'/" {} \;
```

and now am having a separate issue with a missing "bufferutils" in the bundle. Not sure what that is, but looking into that at the moment. If you have any clues, please share!

**Reply by jeffpeck10x** [reply]
For that matter, if you know of a better way to bundle deck.gl in a library or have a working example that would be really really helpful.

**Reply by jeffpeck10x** [reply]
For reference, here is the same issue happening in the latest luma.gl used by deck.gl 8.7:
```
[!] Error: 'global' is not exported by ../../node_modules/probe.gl/env.js, imported by ../../node_modules/@luma.gl/gltools/dist/esm/polyfill/polyfill-vertex-array-object.js
https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module
../../node_modules/@luma.gl/gltools/dist/esm/polyfill/polyfill-vertex-array-object.js (1:9)
```

**Reply by ibgreen** [reply]
Last time we got error reports around rollup we did a push on replacing the troublesome `probe.gl/env` import with a `@probe.gl/env` module and updated the code to using `globalThis` instead of `global` etc.

Looks like luma.gl 8.5-release branch still has some references to `probe.gl/env`, e.g. https://github.com/visgl/luma.gl/blob/8.5-release/modules/gltools/src/polyfill/polyfill-vertex-array-object.js#L29

**Reply by Pessimistress** [reply]
@jeffpeck10x  Can you try removing your lock file and reinstalling the dependencies? I just published luma.gl v8.5.11 which may resolve the issue.

**Reply by jeffpeck10x** [reply]
@Pessimistress I actually gave up on the rollup build, but I _did_ get it working with parcel. There were a few issues, but one of the things that is relevant here is this:
```
#!/usr/bin/env bash

# Cause luma to import from "@probe.gl/env" (with "@") instead of "probe.gl/env"
find ../../node_modules/@luma.gl \
  -name "*.js" \
  -path "*/dist/*" \
  -exec sed -i ""  "s/'probe\.gl\/env'/'@probe.gl\/env'/" {} \;
```
(pasted above as well).

Simply replacing those imports seemed to do the trick for this particular issue.

That you have released an official update to luma.gl that more formally fixes this is great news and I intend to try it out, which would be nice to replace this workaround. Thanks for addressing this!

**Reply by sdeck51** [reply]
Do you have a repo demonstrating it packaging correctly? I've also been trying to bundle deck.gl via rollup for a browser library, using various plugins to get it to bundle correctly, but have ran into similar issues. Haven't used parcel before but if it works I'd gladly switch over to it.

**Reply by jeffpeck10x** [reply]
@sdeck51 sorry I did not respond earlier.

With the latest luma and using parcel, I am able to bundle deck.gl, but with the annoying caveat that things need to be re-exported, like so:
```
import {
  DataFilterExtension as BaseDataFilterExtension,
  PathStyleExtension as BasePathStyleExtension,
} from "@deck.gl/extensions";

export class DataFilterExtension extends BaseDataFilterExtension {}
export class PathStyleExtension extends BasePathStyleExtension {}
```

If I just do:
```
export {
  DataFilterExtension,
  PathStyleExtension,
} from "@deck.gl/extensions";
```

I get this at runtime when I try to use the bundled package:
```
Uncaught ReferenceError: $c0a045143cd4eb0b$re_export$DataFilterExtension is not define
```

This is with the latest parcel 2.6 and seems to be a possible bug in parcel - not related to deck.gl, as it happens with other imports too.

With this small workaround or rexporting things though, I am able to make a bundled package of various gltools, including specific luma and deckgl dependencies.

I should note that in my particular case, I am not including the imported libraries in the bundle (so maybe I shouldn't call it a "bundle"?). That _was_ my original intention, however I ended up not doing that for various reasons. So, maybe the above workaround does not actually apply if the libraries are included?

Either way, I can confirm that this works.
